GenericSerialDeviceRemoteControl/GenericSerialDeviceRemoteControl.py

```diff
@@ -0,0 +1,350 @@
+
+    def onConnectionChanged(self, caller=None, event=None, callData=None):
+        """Call when the connector node is changed, or if the device is (dis)connected."""
+        if self._connectorNode and self._connectorNode.GetState() == slicer.vtkMRMLIGTLConnectorNode.StateConnected:
+            self.onConnected()
+        else:
+            self.onDisconnected()
+
+    def setDeviceID(self, deviceId):
+        """
+        Set the device id from the Plus config file of the associated vtkPlusGenericSerialDevice.
+
+        Input: device id (str, ex. "MySerialDevice")
+        """
+        self._deviceId = deviceId
+
+    def deviceId(self):
+        """
+        Get the currently set Plus config file device id.
+
+        Output: device id (str)
+        """
+        return self._deviceId
+
+    def sendSendTextCommand(self, sendText, callbackOnCommandComplete, blocking=False, timeout_seconds=5.0):
+        """
+        Send a SendText command over IGTL, which sends data over the TX line on the serial device.
+
+        Input: text to send (str), callback function to run when command returns (function)
+        """
+        if self._connectorNode is None or self._connectorNode.GetState() != slicer.vtkMRMLIGTLConnectorNode.StateConnected:
+            logging.error("Connector node unavailable!")
+            if isinstance(self._connectorNode, slicer.vtkMRMLIGTLConnectorNode):
+                logging.error("State: %s", self._connectorNode.GetState())
+            return
+        logging.info("Sending command: %s", sendText)
+        command_xml = f"""<Command Name="SendText" DeviceId="{self._deviceId}" Text="{sendText}" />"""
+        self._command = slicer.vtkSlicerOpenIGTLinkCommand()
+        self._command.SetName("SendText")
+        self._command.SetCommandContent(command_xml)
+        self._command.SetTimeoutSec(timeout_seconds)
+        self._command.SetBlocking(blocking)
+        self._command.ClearResponseMetaData()
+        self.removeObservers(callbackOnCommandComplete)
+        self.addObserver(self._command, slicer.vtkSlicerOpenIGTLinkCommand.CommandCompletedEvent, callbackOnCommandComplete)
+        self._connectorNode.SendCommand(self._command)
+
+    def onSendTextCompleted(self, caller=None, event=None, callData=None):
+        """Sample callback function on a SendText command completion event."""
+        if not self._comand.GetSuccessful():
+            return
+        response = self._command.GetResponseMetaDataElement()
+        logging.info(response)
+
+    def sendGetCTSCommand(self, callbackOnCommandComplete, blocking=False, timeout_seconds=5.0):
+        """
+        Send a GetCTS command over IGTL, which grabs the value from the CTS line on the serial device.
+
+        Input: callback function to run when command returns (function)
+        Output: success (bool)
+        """
+        if self._connectorNode is None or self._connectorNode.GetState() != slicer.vtkMRMLIGTLConnectorNode.StateConnected:
```

Comment:
These couple of lines are repeated 4 times, consider factoring out as much as possible into a function that you call from all 4 places.